Integration of 'aosp-xenvm-trout' AOSP device

svlad-90 opened the pull request:
Integration of 'aosp-xenvm-trout' AOSP device to the open-source

Commit:
Declare 'aosp_xenvm_trout' device

- Added declaration of the 'aosp_xenvm_trout' device
- Added VS Code project files

Signed-off-by: Vladyslav Goncharuk <vladyslav_goncharuk@epam.com>
Reviewed-by: Andrii Chepurnyi <andrii_chepurnyi@epam.com>

Commit:
Override the 'init.vendor.rc' file

We need to prevent renaming of the eth0 network interface on start-up
of the system. This patch introduced init.vendor.rc file, which
comments out the following line:

start rename_eth0

- Introduced 'shared/config/init.vendor.rc', which overrides the
'device/google/cuttlefish/shared/config/init.vendor.rc' file from the
cuttlefish device
- Introduced 'aosp_xenvm_trout_common.mk', which copies
'device/epam/aosp-xenvm-trout/shared/config/init.vendor.rc' as
'/vendor/etc/init/init.cutf_cvm.rc' on the target
- Included 'device/epam/aosp-xenvm-trout/aosp_xenvm_trout_common.mk'
into the 'aosp_xenvm_trout_arm64.mk'
- Included 'device/epam/aosp-xenvm-trout/aosp_xenvm_trout_common.mk'
into the 'aosp_xenvm_trout_x86.mk'

Signed-off-by: Vladyslav Goncharuk <vladyslav_goncharuk@epam.com>
Reviewed-by: Andrii Chepurnyi <andrii_chepurnyi@epam.com>

Commit:
Override of the vehicle@2.0-default-impl-lib

Full library name:
android.hardware.automotive.vehicle@2.0-default-impl-lib

Override is needed to comment out the following 2 lines in the
EmulatedVehicleHal.cpp file:

initObd2LiveFrame(*mPropStore->getConfigOrDie(OBD2_LIVE_FRAME));
initObd2FreezeFrame(*mPropStore->getConfigOrDie(OBD2_FREEZE_FRAME));

Presense of those 2 line is causing a VHAL crash in getConfigOrDie.

For some reason config for above 2 properties is not presented. As we
are not critically interested in them, it was decided to comment them
out.

As we override the library, we should override users of that
library. Those are:

- android.hardware.automotive.vehicle@2.0-virtualization-client-lib
- android.hardware.automotive.vehicle@2.0-virtualization-service

Done items:

- Overriden the
'android.hardware.automotive.vehicle@2.0-default-impl-lib'
- Overriden the
'android.hardware.automotive.vehicle@2.0-virtualization-client-lib'
library
- Overriden the
'android.hardware.automotive.vehicle@2.0-virtualization-service'
service
- Added required sepolicy rules

Signed-off-by: Vladyslav Goncharuk <vladyslav_goncharuk@epam.com>
Reviewed-by: Andrii Chepurnyi <andrii_chepurnyi@epam.com>

Commit:
GPU sepolicy configuration

In order to use virtio GPU we need to set a correct context and provide
additional configurations.

Done items:

- Added sepolicy/genfs_contexts directives, which allow access to a
GPU virtio PCI device. Values are provided for x86 and arm64. They are
compatible with QEMU
- Added GPU-related directives to the sepolicy/sysfs_gpu_file.te, which
allows access to surfaceflinger

Signed-off-by: Vladyslav Goncharuk <vladyslav_goncharuk@epam.com>
Reviewed-by: Andrii Chepurnyi <andrii_chepurnyi@epam.com>

Commit:
Allow usage of the external prebuilt kernel

- Added functionality, which allows usage of the external prebuilt
kernel and modules in the 'xenvm_trout_arm64' device

Signed-off-by: Dmytro Terletskyi <dmytro_terletskyi@epam.com>
Reviewed-by: Andrii Chepurnyi <andrii_chepurnyi@epam.com>

Commit:
Deploy vsock/grpc backends

The idea of the 'aosp-xenvm-trout' device is to be executed on the Xen
hypervisor inside the DomA. This guest domain relies on existance of the
virtio devices, shared from the DomD domain.

But besides those devices the 'aosp-xenvm-trout' relies on the
availability of the several grpc/vsock backends. They are providing data
to the 'Audio Control HAL', 'Dumpstate HAL', 'Vehicle HAL'.

Those backends are packed by Google into the 'agl_services_build' Yocto
meta-layer. It should be built and deployed inside DomD.

The purpose of this patch it to copy and extend the agl_services_build
functionality from the 'trout' device in a way that it could be used
in DomD.

Keep in mind, that until vsock communication is not proved to work, we
leave local grpc server inside Android up and running. So it will be
running both in DomD and inside Android. That should do no harm as there
is only one client which will connect only to one of the available
sources.

Signed-off-by: Vladyslav Goncharuk <vladyslav_goncharuk@epam.com>
Reviewed-by: Andrii Chepurnyi <andrii_chepurnyi@epam.com>

Commit:
Switch to eng. build

For virtio research we are using quite a new Android kernel
( common-android14-6.1 ), which does not pass runtime checks on
Android 12, when non-eng. build is used.

Thus, we are switching to the eng. build, where such runtime checks are
skipped.

Signed-off-by: Vladyslav Goncharuk <vladyslav_goncharuk@epam.com>
Reviewed-by: Andrii Chepurnyi <andrii_chepurnyi@epam.com>

Commit:
Extend rootfs size to 2Gb

During execution of CTS it was identified that there is not enough space
in root file-system to copy all the necessary files. This patch is
intended to extend the size of root fs from 1 to 2 Gb.

Signed-off-by: Vladyslav Goncharuk <vladyslav_goncharuk@epam.com>
Reviewed-by: Andrii Chepurnyi <andrii_chepurnyi@epam.com>

Commit:
Adjusted path to PCI device in the genfs_contexts

On the physical board we have a different PCI bus identifier, that the
one, which we were using for QEMU. Thus, we need to adjust sepolicy
rules to allow usage of that device by the graphics sub-system.

Signed-off-by: Vladyslav Goncharuk <vladyslav_goncharuk@epam.com>
Reviewed-by: Andrii Chepurnyi <andrii_chepurnyi@epam.com>

arminn commented on shared/config/init.vendor.rc:
```diff
@@ -0,0 +1,152 @@
+on early-init
```
Comment to patch "Override the 'init.vendor.rc' file" : Please add into description why we need to override original init.vendor.rc

svlad-90 replied:
Hi @arminn,

I've checked why this change was done. It was introduced in order to prevent the renaming of the eth0 interface during the startup of the system. I've extended the commit message.

Done.

arminn commented on aosp_xenvm_trout_arm64.mk:
```diff
@@ -2,6 +2,13 @@ $(call inherit-product, device/google/trout/aosp_trout_arm64.mk)
 
 include device/epam/aosp-xenvm-trout/aosp_xenvm_trout_common.mk
 
+DEVICE_MANIFEST_FILE := device/google/trout/manifest_x86.xml
```
Comment to patch "Override trout VHAL implementation" : The patch "Override trout VHAL implementation" should be reworked and divided  into at least 3  parts since there are several not connected changes. Also additional information regarding why audio.primary.caremu overrided should be added

svlad-90 replied:
The change was split into 3 commits:

https://github.com/xen-troops/android_device_epam_xenvm-trout/pull/1/commits/397fd4d1f4b8672a5aeac3dd015afe0429d032bb

https://github.com/xen-troops/android_device_epam_xenvm-trout/pull/1/commits/647156544e7a19d2a8949886d6c0fec5a23b84c8

https://github.com/xen-troops/android_device_epam_xenvm-trout/pull/1/commits/50e2fab5be43eb7e5f978a10a81e12b08ccb3b7f

The sum of all 3 changes is equal to the previous bif commit, except that I've removed the following 3 lines:

```
DEVICE_MANIFEST_FILE := device/google/trout/manifest_x86.xml
DEVICE_MATRIX_FILE += device/google/trout/compatibility_matrix.xml
DEVICE_FRAMEWORK_COMPATIBILITY_MATRIX_FILE := device/google/trout/framework_compatibility_matrix.xml
```

I don't remember why they were needed and was not able to explain the motivation. It seems that they are not needed for arm64. After all changes for all your comments will be done, I'll do a testing session. If these 3 lines would have a critical meaning for having the system up and running. I'll return them back.

Done.

arminn commented on aosp_xenvm_trout_common.mk:
```diff
@@ -7,5 +7,10 @@ PRODUCT_COPY_FILES += \
 LOCAL_VHAL_PRODUCT_PACKAGE ?= android.hardware.automotive.vehicle@2.0-virtualization-service-extended
 
-PRODUCT_PACKAGES += audio.primary.caremu.extended \
+LOCAL_AUDIO_PROPERTIES ?= \
```
Comment on patch "Override implementation of emulator's audio driver" : I believe this patch should be a solid part with the audio code from the "Override trout VHAL implementation"

svlad-90 replied:
Hi @arminn, this patch was moved to the following commits:

https://github.com/xen-troops/android_device_epam_xenvm-trout/pull/1/commits/50e2fab5be43eb7e5f978a10a81e12b08ccb3b7f

https://github.com/xen-troops/android_device_epam_xenvm-trout/pull/1/commits/b94032ae0a0317d205ede8e4b34fbcf112a77177 ( there was one last line in this commit related to vehicle topics )

This patch was removed, as it is empty now.

Done.

arminn commented on sepolicy/genfs_contexts:
```diff
@@ -0,0 +1,5 @@
+genfscon sysfs /devices/pci0000:00/0000:00:04.0 u:object_r:sysfs_gpu_trout:s0
+
+# arm64
+genfscon sysfs /devices/platform/4010000000.pcie/pci0000:00/0000:00:03.0 u:object_r:sysfs_gpu_trout:s0
```
Comment on patch "Extended sepolicy rules" : It looks like patch sets the correct context for virtio gpu and and provides additional configuration for access. Please update description with respect to it.

svlad-90 replied:
Hi @arminn,

Done.

arminn commented on sepolicy/genfs_contexts:
```diff
@@ -2,4 +2,4 @@
 
 # arm64
-genfscon sysfs /devices/platform/4010000000.pcie/pci0000:00/0000:00:03.0 u:object_r:sysfs_gpu_trout:s0
+genfscon sysfs /devices/platform/10000000.pcie/pci0000:00/0000:00:03.0 u:object_r:sysfs_gpu_trout:s0
```
Comment on patch "Adjusted path to PCI device in the genfs_contexts" : this should be integrated into "Extended sepolicy rules" including description

svlad-90 replied:
Hi @arminn, the previous value was compatible with QEMU. While the new one is compatible with the physical device.

I think we want to have both those values in our commit history in order to be able to "remember" them, Someone might want to use QEMU. Then he might fetch the previous value form the history.

Let me know, whether you agree with this.

arminn replied:
Ok make sense

arminn commented on aosp_xenvm_trout_common.mk:
```diff
@@ -5,5 +5,7 @@ PRODUCT_COPY_FILES += \
     device/epam/aosp-xenvm-trout/shared/config/init.recovery.rc:$(TARGET_COPY_OUT_RECOVERY)/root/init.recovery.xenvm_trout.rc \
     device/epam/aosp-xenvm-trout/shared/config/ueventd.rc:$(TARGET_COPY_OUT_RECOVERY)/root/ueventd.xenvm_trout.rc
 
+LOCAL_VHAL_PRODUCT_PACKAGE ?= android.hardware.automotive.vehicle@2.0-virtualization-service-extended
```
Comment to patch "Override cuttlefish VHAL implementation" : Please provide information in description why this override needed

svlad-90 replied:
Hi @arminn,

New commit: https://github.com/xen-troops/android_device_epam_xenvm-trout/pull/1/commits/a38d2538259ba6467c7f6dfcd0fed6365d65316b

I've extended its description.

In short - all this is done to comment these 2 lines:

```
initObd2LiveFrame(*mPropStore->getConfigOrDie(OBD2_LIVE_FRAME));
initObd2FreezeFrame(*mPropStore->getConfigOrDie(OBD2_FREEZE_FRAME));
```

A more detailed explanation is in the commit message.

Done.

arminn commented on sepolicy/file_contexts:
```diff
@@ -1,4 +1,3 @@
-
```
Comment on patch "Override implementation of emulator's audio driver" : this change not needed.

svlad-90 replied:
Hi @arminn,

Done.

arminn commented on aosp_xenvm_trout_arm64.mk:
```diff
@@ -45,7 +45,7 @@ include device/epam/aosp-xenvm-trout/aosp_xenvm_trout_common.mk
 DEVICE_MANIFEST_FILE := device/google/trout/manifest_x86.xml
 DEVICE_MATRIX_FILE += device/google/trout/compatibility_matrix.xml
 DEVICE_FRAMEWORK_COMPATIBILITY_MATRIX_FILE := device/google/trout/framework_compatibility_matrix.xml
- 
```
Comment on patch "Extend rootfs size to 2Gb" : this change not needed.

svlad-90 replied:
Hi @arminn, it is hard to follow this comment after some work is done ( GitHub screws it up ). But it seems to be removed now.

Done.

arminn commented:
The patches "Override cuttlefish VHAL implementation" "Deploy vsock/grpc backends" "Revert deletion of overriden_impl" : adding some stuff, removing it and adding again. This should be reworked.

arminn commented:
Note: I did per commit review, for all code comments there is a comment which patch was addressed(github very strangely displays all of that).

Commit:
Introduction of the 'xenvm_trout' hardware

It is a way to have a more deep definition of our own device.

We want to use "androidboot.hardware=xenvm_trout" kernel flag. That
would be a proper way to have our own config *.rc files being searched
during the startup. One of them will avoid renaming of eth0 during
startup of the system.

- Introduced 'shared/config/ueventd.rc', which overrides
'device/google/cuttlefish/shared/config/ueventd.rc'
- Introduced 'shared/config/init.recovery.rc', which overrides
'device/google/cuttlefish/shared/config/init.recovery.rc'
- Extended 'aosp_xenvm_trout_common.mk' with copying 3 specific rc
files to the vendor image with 'xenvm_trout' pattern in their names
- Added 'vendor.ser.bt-uart' parameter to the
'aosp_xenvm_trout_arm64.mk'. Usage of '/dev/hvc5' is declared in
'ueventd.rc'. Without this declaration there were issues with startup.

Signed-off-by: Vladyslav Goncharuk <vladyslav_goncharuk@epam.com>
Reviewed-by: Andrii Chepurnyi <andrii_chepurnyi@epam.com>

Commit:
Override impl. of the virtualization-grpc-server

In order to have 'trout' running we need to have virtualization GRPC
server available. Vehicle hal relies on its presence. In the future, it
should be available in DomD. But as far as we are experimenting with
QEMU, we will deploy slightly extended ( gets more parameters from rc
file ) version of it right inside Android.

The following things are done:

- Added minimalistic override of the virtualization-grpc-server (
android.hardware.automotive.vehicle@2.0-virtualization-grpc-server-extended
) with modified list of parameters and changed service name
- Added related sepolicy rules
- Added service to the PRODUCT_PACKAGES

Signed-off-by: Vladyslav Goncharuk <vladyslav_goncharuk@epam.com>
Reviewed-by: Andrii Chepurnyi <andrii_chepurnyi@epam.com>

Commit:
Added overriden impl. of the audio.primary.caremu

There is a crash in Android 12 version of the "audio.primary.caremu"
service implementation.

This patch introduces the overriden implementation of this service,
which fixes the crash.

The fix of the crash is located in the 'audio_hw.c' line 1221.
pthread_create call should be moved to a certain place in order to avoid
a race condition.

Initially the call was located here:

https://cs.android.com/android/platform/superproject/+/android-
12.1.0_r8:device/generic/car/emulator/audio/driver/audio_hw.c;l=1196

Signed-off-by: Vladyslav Goncharuk <vladyslav_goncharuk@epam.com>
Reviewed-by: Andrii Chepurnyi <andrii_chepurnyi@epam.com>

svlad-90 commented:
Hi @arminn, I've lost one of your comments in GitHub. The one, related to squashing all those "remove, and then add again the overriden_impl". So firing another stand-alone comment - it is done. The redundant patch was removed.

svlad-90 commented:
Hi @arminn, seems that all items are done. I'll spend tomorrow in order to check whether the final result is still a working piece of SW. ))

svlad-90 commented:
Hi @arminn, it was checked that the Android start-up is OK after applying all the refactored source code.

arminn commented:
@svlad-90 add plz after your signoff to patches:

Reviewed-by: Andrii Chepurnyi \<andrii_chepurnyi@epam.com\>

and I will merge PR

svlad-90 commented:
Hi @arminn, I've added the reviewed-by tag to all commits. But, please, do not merge it yet. I want to add a fix for the issue with multiple displays to this PR. Once it's ready - I'll let you know with a separate comment.

Commit:
Remove cluster display port

We've started to observe that for some teammates the content of the
cluster display port was drawn on top of the main display port's
content. To step over this issue we've decided to turn off the cluster
display port.

This patch introduces an overlay, which overrides the car service's
'config_occupant_display_mapping' configuration setting.

Signed-off-by: Vladyslav Goncharuk <vladyslav_goncharuk@epam.com>
Reviewed-by: Andrii Chepurnyi <andrii_chepurnyi@epam.com>

svlad-90 commented:
Hi @arminn, ready for merge.
